### ISSUE TITLE
Scalable GP Classifier on structured feature set, (multi-dim features over several classes)

### ISSUE THREAD

**Issue by twkillian** [cornellius-gp/gpytorch#546]
I've been trying to implement a custom DKL type model for classification purposes. In my case, the feature extractor provides a structured set of features for each possible class, a tensor of (batch_size x N_c x K  ). In the particular problem, there are 10 classes (N_c = 10) with the feature extractor providing 16 features per class (K =10). 

I've been trying to construct the GP classifying layer to receive the feature vector input for each class and then provide a distribution over the final classification decision.

Which scalable GP approach is best suited in this case?

I have been trying to adapt the [DKL example on CIFAR](https://github.com/cornellius-gp/gpytorch/blob/master/examples/08_Deep_Kernel_Learning/Deep_Kernel_Learning_DenseNet_CIFAR_Tutorial.ipynb) while incorporating some of the elements from the [KISS-GP for scalable classification example](https://github.com/cornellius-gp/gpytorch/blob/master/examples/07_Scalable_GP_Classification_Multidimensional/KISSGP_Additive_Classification_CUDA.ipynb).

Here is how I've been trying to write the GP classifier (num_dim and grid_size were chosen to try and knock down the memory requirements... I'm really struggling to gain an intuition about how to properly set these):
```python
class GaussianProcessClassifier(AbstractVariationalGP):
    def __init__(self, num_dim=10, grid_bounds=(-1,1), grid_size=5):
        variational_distribution = CholeskyVariationalDistribution(num_inducing_points=grid_size)
        variational_strategy = AdditiveGridInterpolationVariationalStrategy(self,
                                                                            grid_size=grid_size,
                                                                            grid_bounds=[grid_bounds]*16,
                                                                            num_dim=16,
                                                                            variational_distribution=variational_distribution)
        super(GaussianProcessClassifier,self).__init__(variational_strategy)

        self.covar_module = ScaleKernel(
                                RBFKernel(
                                lengthscale_prior = SmoothedBoxPrior(
                                    math.exp(-1), math.exp(1), sigma=0.1, transform=torch.exp
                                )
                            )
                        )
        self.mean_module = ConstantMean()
        self.grid_bounds = grid_bounds

    def forward(self,x):
        mean = self.mean_module(x)
        covar = self.covar_module(x)
        return MultivariateNormal(mean,covar)
```

So all together, the model looks something like this:
```python
class KernelizedClassifier(gpytorch.Module):
    def __init__(self, feature_extractor, num_classes=10, grid_bounds=(-1,1) ):
        super(KernelizedClassifier,self).__init__()
        self.feature_extractor = feature_extractor
        self.gp_clf = GaussianProcessClassifier(num_dim=num_classes, grid_bounds=grid_bounds)
        self.num_classes = num_classes
        self.grid_bounds = grid_bounds

    def forward(self,x):
        features = self.feature_extractor(x).squeeze()
        features = gpytorch.utils.grid.scale_to_bounds(features,self.grid_bounds[0],self.grid_bounds[1])
        res = self.gp_clf(features)
        return res
```
with the model and likelihood being initialized as:
```python
model=KernelizedClassifier(feature_extractor, num_classes=num_classes).to(device)
likelihood = SoftmaxLikelihood(num_features=num_classes,n_classes=num_classes).to(device)
```
Thanks, any input or guidance would be greatly appreciated.
--Taylor

**Comment by jacobrgardner** [comment]
Hi @twkillian,

At present we recommend the use of the new `WhitenedVariationalStrategy` over `AdditiveGridInterpolationVariationalStrategy` for large numbers of features for VI with SV-DKL, which you can view an example of [here](https://github.com/cornellius-gp/gpytorch/blob/master/examples/05_Scalable_GP_Regression_Multidimensional/SVDKL_Regression_SVGP_CUDA.ipynb) that you should be able to simply swap the `GaussianLikelihood` out for a `SoftmaxLikelihood`. 

With that said, I believe the main issue you are running in to is that at the moment GPyTorch always interprets the last two dimensions as the "number of data points" and "number of features" dimensions respectively. By passing `X` as a `batch_size x N_c x K`, you are basically telling us you want to train `batch_size` independent GPs, each with a minibatch of `N_c` points in `K` dimensions. 

Most likely what you are wanting to do is combine the `N_c` and `K` dimension so that `X` will be a `batch_size x N_c * K` tensor (e.g., `X = X.view(batch_size, -1)`, and `y` is correspondingly a `batch_size` tensor that takes on `N_c` possibly different values. Does this sound like what you are trying to do?

**Comment by twkillian** [comment]
This is very instructive. Thank you.

I'm not entirely sure that we want to combine the number of classes and the set of features constructed per class. What I was hoping to do is have ```N_c``` independent GPs that would take the ```K``` dimensional feature vector associated with that class as input. These feature vectors should be kept more or less independent from one another in determining the final class of the provided input. So, when you say that:

> GPyTorch always interprets the last two dimensions as the "number of data points" and "number of features" dimensions respectively.

Would this mean that I want to pass ```X``` as ```N_c x batch_size x K``` ? As far as I understand, would this correspond to ```N_c``` independent GPs with a minibatch of ```batch_size``` points each with ```K``` dimensions? 

Continuing with this, I would then construct the Likelihood as, ```SoftmaxLikelihood(num_features=K, n_classes=num_classes)```. 

> (Side note, 👎 at SoftmaxLikelihood having both num_features and n_classes arguments!)

You win some, you lose some 🤷‍♂️

**Comment by jacobrgardner** [comment]
Yup, you've got it for the size of X if you want to do that!

You'll also need `y` to be `N_c x batch_size` since completely independent GPs operate on independent outputs. The simplest solution might be to simply repeat `y` to the right size, but what exactly you want will depend on the exact semantics you'd like your model to have.

**Comment by twkillian** [comment]
Okay, I'll play with this a little to make sure that I get the model working the way I want. 

The intent (following SV-DKL) is to combine the GPs to provide a classification over the ```N_c``` classes. Something like ```preds  = argmax(output.probs)``` when using the fully trained model.

**Comment by jacobrgardner** [comment]
In that case combining the two dimensions may still be the way to go. The softmax likelihood includes a mixing weight matrix W that will map from the N_c*K extracted features to N_c outputs.

Alternatively you could mix the K features of the N_c independent GPs together, which might require a slightly customized likelihood. Either way, you're going to just want to turn that GP layer in to something batch_size x N_c so that you can pass it through a softmax or something similar

**Comment by twkillian** [comment]
That makes sense. I'll fix my code and try combining the extracted feature dimensions to get started. 

Thanks for being so helpful and prompt.

**Comment by twkillian** [comment]
@jacobrgardner Thanks again for your help this past weekend. It's unclear whether ```WhitenedVariationalStrategy``` is appropriate for Classification purposes (at least to the point that it won't necessarily work with ```SoftmaxLikelihood```) since it marginalizes over the features/inducing points. This is problematic when the Likelihoods used for classification (as you mentioned above) learn a mixing weight matrix W to help predict classes.

From what I have gathered, ```SoftmaxLikelihood``` expects the output posterior MVN to have the internal structure of 'number of features' x 'number of data points/independent GPs'. ```WhitenedVariationalStrategy``` integrates over that first dimension causing an error inside ```SoftmaxLikelihood``` since the samples from the latent function are missing that dimension ([source, for reference (lines 34-42)](https://github.com/cornellius-gp/gpytorch/blob/adc4c36e547cf8da42c4cc653a68e28d74b69c25/gpytorch/likelihoods/softmax_likelihood.py#L34))

Rather than open a new issue since I'm still trying to fully understand how everything works... I just thought that I'd update the issue.

What I chose to do in the interim is follow your suggestion of collapsing the number of classes (```N_c```) and feature (```K```) dimensions and chose to inherit ```gpytorch.models.AdditiveGridInducingVariationalGP``` in my GP Classifying layer, following the [DKL example](https://github.com/cornellius-gp/gpytorch/blob/adc4c36e547cf8da42c4cc653a68e28d74b69c25/examples/08_Deep_Kernel_Learning/Deep_Kernel_Learning_DenseNet_CIFAR_Tutorial.ipynb). I'm not completely satisfied with how the semantics of my model play out but at present, this approach appears to be working. I'll keep hacking away at this, trying to build a deeper understanding of how I want this whole thing to work out.
